src/nagini_translation/translators/call.py

```diff
@@ -149,6 +149,35 @@ def _translate_super(self, node: ast.Call, ctx: Context) -> StmtsAndExpr:
         else:
             raise InvalidProgramException(node, 'invalid.super.call')
 
+    def translate_adt_cons(self, cons: PythonClass, args: List, pos,
+                           ctx: Context) -> Expr:
+        """
+        Constructs ADTs via a sequence of constructor calls and
+        boxing/unboxing calls.
+        """
+        info = self.no_info(ctx)
+        adt_name = cons.adt_def.name + '_ADT'
+        adt_prefix = cons.adt_def.name + '_'
+        adt_type = self.viper.DomainType(adt_name, {}, [])
+
+        # If expected argument type is the ADT type (another constructor call),
+        # unbox translated argument
+        for index, ((_, arg_type), translated_arg) in enumerate(zip(cons.fields.items(),
+                                                                args)):
+            if arg_type.type == cons.adt_def:
```

Comment:
I don't think I understand this. Why do we need special treatment if the argument type is also the type of this ADT? If we need special treatment for this, does that mean something wouldn't work if the argument type was a different ADT type?

Reply:
If the type of the argument is the ADT type it is treated natively, i.e., we unbox it to match the signature of the function. Any other type is treated as Ref, including other ADTs. An alternative would be to treat everything as Ref: both "foreign" types and the ADT itself. With the proper boxing and unboxing, the same level of type verification could be reached. I chose this approach for three reasons: it uses Viper's own type system (providing more objective error messages), it can be faster in this way (as type checking in the AST can be faster than axiomatizing types and calling the SMT solver, though equivalent) and it is the reference ADT encoding in Viper tutorial.